### PR TITLE
fix: Make regen less greedy

### LINKED ISSUE #3021: group and chord: subtasks are not executed until entire generator is consumed
When passing a generator to `group` or `chord`, sub-tasks are not submitted to the broker for execution until the entire generator is consumed.  The expected result is that subtasks are submitted during iteration, as soon as they are visited.

Here is the test code I'm using:

``` python
from celery import Celery, group
from celery.utils.log import get_task_logger

app = Celery(__name__, broker='redis://localhost:6379/0',
             backend='redis://localhost:6379/0')

logger = get_task_logger(__name__)

def longyield():
    import time
    for i in xrange(5):
        time.sleep(2)
        print "yielding", i
        yield add.s(i, i)

@app.task
def add(x, y):
    logger.info("adding {} + {}".format(x, y))
    return x + y

def itertest():
    it = longyield()
    g = group(it)
    result = g.delay()
    print result.get()
```

I'm testing with 4.0rc1.

### LINKED ISSUE #4298: group.skew() removes all tasks if group input is generator
## Checklist

- [X] I have included the output of ``celery -A proj report`` in the issue.
      (if you are not able to do this, then at least specify the Celery
       version affected).

      python 3.6.2
      celery==4.1.0
      amqp==2.2.1
      billiard==3.5.0.3

- [X] I have verified that the issue exists against the `master` branch of Celery.
      As of `be55de6`

## Steps to reproduce
```
@celery.task(bind=True)
def noop(task):
    pass
```

1. `group = celery.group(noop.si() for i in range(10))`
2. `group.skew()`  -> `group(<empty>)`
3. `group.apply_async()` or `group()`  -> `<GroupResult: 66fd517e-123b-4299-8cbe-9183b3f02626 []>`

## Expected behavior
```
<GroupResult: d610c923-e939-4199-9bb7-7bc89daa2ccb [9152bd60-f815-4bbe-a407-a003db34b19d, 7fc2d0ac-72ab-48b1-9083-be0f3b50b00d, d7c0a00a-85c3-4780-88f9-23be8da6fb73, 3737944f-91a4-4886-95c6-4960ef4764a7, a8da96ea-d2c4-4876-8026-e0c24c7d508a, 83ee5316-fb2f-472b-8e94-284144f92438, 5715a751-f28d-4deb-ac15-114854570fde, bc1646bf-cec9-4740-8e74-411d7b66dcca, 337aa1f3-5606-45b2-af68-5c5527289019, f5fffb98-f2f5-4423-a656-8c1d7e93291a]>
```

## Actual behavior

```
<GroupResult: 66fd517e-123b-4299-8cbe-9183b3f02626 []>
```

### Notes
If you `print(group)`(or something that evaluates `self.tasks`) before calling `group.skew()`, it works normally.

### PR DESCRIPTION
**Pull request by maybe-sybr**
## Description

This change is pulled from #3043 (and was also included in modified form in #4459 and #6112). It is the body of change to the `functional.regen` util which is sometimes greedy and also sometimes lost items per the issue reported in #4298.

This change fixes #4298 and we'll need to revisit the diffs from the other tagged PRs if we intend to address #3021.

### CODE REVIEW

**Comment by maybe-sybr** [comment]
I've pulled and uplifted the tests added for `regen` in the original PR (#3043), but will need to check coverage prior to merge.

**Comment by maybe-sybr** [comment]
@auvipy - I know #4459 was from quite a long time ago but I'd appreciate a review from you since you've seen these PRs before :)

**Comment by maybe-sybr** [comment]
Huh, where did Travis go in the checks? @thedrow any ideas?

**Comment by thedrow** [comment]
Travis no longer works since we're out of credit.

File: celery/utils/functional.py [hunk 1]
```diff
@@ -182,38 +182,73 @@ def __init__(self, it):
...
+        if self.__done:
+            return repr(self.__consumed)
+        else:
+            return '[{0}]'.format(', '.join(
+                [repr(x) for x in self.__consumed] + ['...'])
+            )
+
     def __iter__(self):
-        return chain(self.__consumed, self.__it)
+        for x in self.__consumed:
```
Comment (thedrow) [inline celery/utils/functional.py:203]: `yield from ` is shorter...

File: celery/utils/functional.py [hunk 2]
```diff
@@ -182,38 +182,73 @@ def __init__(self, it):
...
         try:
-            self.__consumed.extend(list(self.__it))
+            next(iter(self))
         except StopIteration:
-            pass
+            return False
+        else:
+            return True
+
+    # Python3
```
[the remark this answers is not in the document]
Comment (maybe-sybr) [reply celery/utils/functional.py:238]: I think not, we should just rename the `__nonzero__()` definition above to `__bool__()` now that we don't need to worry about 2.7. Good pick up.

File: celery/utils/functional.py (hunk 1 above)
Comment (maybe-sybr) [reply celery/utils/functional.py:203]: Yep, I'll change this and have a more critical read through for other 2.7isms which could be removed from this diff.

**Comment by maybe-sybr** [comment]
/close
Closing in favour of #6589 which landed earlier

**Pull request closed without merging**